File: process-scala/src/main/scala-2.11/Process.scala
import java.io.{File, IOException}

import com.github.tototoshi.csv.CSVWriter
import org.json4s.JsonAST._
import org.json4s.native.JsonMethods._

import scala.collection.mutable
import scala.io.Source

object Process {

  def main(args: Array[String]): Unit = for (_ <- 1 to 100) { process("Name", "../test_data/") }

  def process(field: String, folder: String): String = {
    val frequencies = mutable.Map.empty[String, Int]

    val files = new File(folder).listFiles().filter(x => x.isFile && x.getName.endsWith(".json"))
    files foreach { file =>
      val contents = Source.fromFile(file).mkString
      val json = parse(contents)
      json \ field match {
        case JString("") =>
        case JString(value) => frequencies(value) = frequencies.getOrElse(value, 0) + 1
        case JNothing => throw new IOException("Field is missing")
        case _ => throw new IOException("Field is not a string")
      }
    }
    val sorted = frequencies.toList.sortWith(_._2 > _._2)

    val outfile = new File("output.csv")
    val writer = CSVWriter.open(outfile)
    writer.writeAll(sorted.map(x => Seq(x._1, x._2.toString)))
    writer.close()

    "ok"
  }
}

File: process-kotlin/src/main/kotlin/process.kt
@file:JvmName("Process")
package process

import kotlin.system.exitProcess
import java.lang.System
import java.lang.Exception
import java.io.IOException
import java.io.File
import com.fasterxml.jackson.dataformat.csv.CsvParser
import com.fasterxml.jackson.dataformat.csv.CsvMapper
import com.fasterxml.jackson.databind.ObjectMapper

val OK   = 0
val FAIL = 1

fun printerr(message: Any?) = System.err.println(message)

fun process(field: String, folder: String): Int {
    val histogram = mutableMapOf<String, Int>()

    val files = File(folder).listFiles { file -> file.isFile() && file.getName().endsWith(".json") }
    val jsonMapper = ObjectMapper()
    files?.forEach { file ->
        val data: Map<*, *>
        try {
            data = jsonMapper.readValue(file, Map::class.java)
        } catch (ex: Exception) {
            printerr(ex)
            return FAIL
        }
        val key = data.get(field)
        when (key) {
            null -> {
                printerr("Field is missing")
                return FAIL
            }
            "" -> {}
            is String -> {
                var count = histogram.get(key)?.plus(1) ?: 1
                histogram.put(key, count)
            }
            else -> {
                printerr("Field is not a string")
                return FAIL
            }
        }
    }
    val frequencies = histogram.asSequence().sortedBy {
        kv -> -kv.value
    }.map {
        kv -> arrayOf(kv.key, "${kv.value}")
    }.toList()
    return writeCSV("output.csv", frequencies)
}

fun writeCSV(filename: String, data: List<*>): Int {
    val mapper = CsvMapper()
    mapper.enable(CsvParser.Feature.WRAP_AS_ARRAY)
    val outFile = File(filename)
    try {
        mapper.writeValue(outFile, data)
    } catch (e: IOException) {
        printerr(e)
        return FAIL
    }
    return OK
}

fun main(args: Array<String>) {
    if (2 != args.size) {
        printerr("Args are: <field name> <folder>")
        exitProcess(FAIL)
    } else {
        exitProcess(process(args[0], args[1]))
    }
}

//// Benchmark
//fun main(args: Array<String>) {
//    for (x in 1..100)
//        process("Name", "../test_data/")
//}
